plugins/notifications-backend/src/service/router.ts

```diff
@@ -1,61 +1,160 @@
+          const parts = relation.targetRef.split(':');
+          if (parts.length == 0 || parts[0] != 'group') {
+            return;
+          }
+          myRes = `${myRes}:${parts[1]}`;
+        });
+      }
+    }
+    
+    response.json({ status: myRes });
   });
 
-  router.get('/notifications/count', async (_, response) => {
+  router.get('/notifications/count/:text?', async ({params: {text}}, response) => {
     let msgcount = -1;
-    await dbClient('messages')
-      .count('* as CNT')
-      .then(count => {
+    
+    var query = dbClient('messages').count('* as CNT');
+    
+    if ( text && text.length > 0 ) {
+      query.whereILike('subject', `%${text}%`).orWhereILike('body', `%${text}%`)
+    }
+    
+    await query.then(count => {
         msgcount = (count[0]?.CNT as unknown as number) || -1;
       });
+
     response.json({ status: 'ok', count: msgcount });
   });
 
   router.post('/notifications', async (request, response) => {
     let msgid = -1;
+
+    var createRequest = request.body as CreateRequest;
+    
+    if (createRequest.targetGroups) {
+      createRequest.targetGroups.forEach(async group => {
+        const groupRef = await catalogClient.getEntityByRef("group:"+group);
+        if ( !groupRef ) {
+          response.json({ status: 'fail', error: 'group does not exist: '+group });
+        }
+      });
+    }
+
+    if (createRequest.targetUsers) {
+      createRequest.targetUsers.forEach(async user => {
+        const userRef = await catalogClient.getEntityByRef("user:"+user);
+        if ( !userRef ) {
+          response.json({ status: 'fail', error: 'user does not exist: '+user });
+        }
+      });
+    }
+
+    const dbObj = {subject: createRequest.subject, body: createRequest.body};
+
     await dbClient('messages')
-      .insert(request.body)
+      .insert(dbObj)
       .returning('id')
       .then(id => {
         // We should harden the type
         msgid = id as unknown as number;
       });
     response.json({ status: 'ok', msgid: msgid });
   });
-  router.get('/notifications', async (_, response) => {
+  
+  router.get('/notifications/:text?/:pageSize?/:pageNumber?', async ({params: {text, pageSize, pageNumber}}, response) => {
     let messages = '{}';
-    await dbClient('messages')
-      .select('*')
-      .then(body => {
+    
+    var pageSizeNum = 0
+    var pageNumberNum = 0
+    var offset = 0
+
+    if ( pageSize && pageNumber ) {
+      pageSizeNum = Number.parseInt(pageSize)
+      pageNumberNum = Number.parseInt(pageNumber)
+      offset = pageSizeNum * (pageNumberNum - 1)
+    }
+    
+    var query = dbClient('messages').select('*');
+
+    if ( text && text.length > 0 ) {
+      query.whereILike('subject', `%${text}%`).orWhereILike('body', `%${text}%`)
+    }
+
+    if ( pageNumberNum != 0 ) {
+      query.limit(pageNumberNum).offset(offset);
+    }
+    
+    await query.then(body => {
         messages = body as unknown as string;
```

Comment:
I am for naming the requested DB columns in the SELECT, typing them and composing the result explicitly instead of blind passing the result.

Reply:
of course. we have to export a type for the returned body